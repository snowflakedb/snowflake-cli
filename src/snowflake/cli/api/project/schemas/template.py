# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from typing import Any, List, Literal, Optional, Union

import typer
from click import ClickException
from pydantic import BaseModel, ConfigDict, Field
from snowflake.cli.api.exceptions import InvalidTemplate
from snowflake.cli.api.secure_path import SecurePath


class TemplateVariable(BaseModel):
    model_config = ConfigDict(frozen=True)

    name: str = Field(..., title="Variable identifier")
    type: Optional[Literal["string", "float", "int"]] = Field(  # noqa: A003
        title="Type of the variable", default=None
    )
    prompt: Optional[str] = Field(title="Prompt message for the variable", default=None)
    default: Optional[Any] = Field(title="Default value of the variable", default=None)

    @property
    def python_type(self):
        # override "unchecked type" (None) with 'str', as Typer deduces type from the value of 'default'
        return {
            "string": str,
            "float": float,
            "int": int,
            None: str,
        }[self.type]

    def prompt_user_for_value(self, no_interactive: bool) -> Union[str, float, int]:
        if no_interactive:
            if not self.default:
                raise ClickException(f"Cannot determine value of variable {self.name}")
            return self.default

        prompt = self.prompt if self.prompt else self.name
        return typer.prompt(prompt, default=self.default, type=self.python_type)


class Template(BaseModel):
    minimum_cli_version: Optional[str] = Field(
        None, title="Minimum version of Snowflake CLI supporting this template"
    )
    files_to_render: List[str] = Field(title="List of files to be rendered", default=[])
    variables: List[TemplateVariable] = Field(
        title="List of variables to be rendered", default=[]
    )

    def __init__(self, template_root: SecurePath, **kwargs):
        super().__init__(**kwargs)
        self._validate_files_exist(template_root)

    def merge(self, other: Template):
        if not isinstance(other, Template):
            raise ClickException(f"Can not merge template with {type(other)}")

        errors = []
        if self.minimum_cli_version != other.minimum_cli_version:
            errors.append(
                f"minimum_cli_versions do not match: {self.minimum_cli_version} != {other.minimum_cli_version}"
            )
        variable_map = {variable.name: variable for variable in self.variables}
        for other_variable in other.variables:
            if self_variable := variable_map.get(other_variable.name):
                for attr in ["type", "prompt", "default"]:
                    if getattr(self_variable, attr) != getattr(other_variable, attr):
                        errors.append(
                            f"Conflicting variable definitions: '{self_variable.name}' has different values for attribute '{attr}': '{getattr(self_variable, attr)}' != '{getattr(other_variable, attr)}'"
                        )
        if errors:
            error_str = "\n\t" + "\n\t".join(error for error in errors)
            raise ClickException(
                f"Could not merge templates. Following errors found:{error_str}"
            )
        self.files_to_render = list(set(self.files_to_render + other.files_to_render))
        self.variables = list(set(self.variables + other.variables))
        return self

    def _validate_files_exist(self, template_root: SecurePath) -> None:
        for path_in_template in self.files_to_render:
            full_path = template_root / path_in_template
            if not full_path.exists():
                raise InvalidTemplate(
                    f"[files_to_render] contains not-existing file: {path_in_template}"
                )
            if full_path.is_dir():
                raise InvalidTemplate(
                    f"[files_to_render] contains a dictionary: {path_in_template}"
                )
