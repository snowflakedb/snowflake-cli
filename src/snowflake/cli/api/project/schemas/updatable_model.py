# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from contextlib import contextmanager
from contextvars import ContextVar
from typing import Any, Dict, Iterator, Optional

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    ValidationInfo,
    field_validator,
)
from pydantic.fields import FieldInfo
from snowflake.cli.api.project.util import IDENTIFIER_NO_LENGTH

PROJECT_TEMPLATE_START = "<%"


def _is_templated(info: ValidationInfo, value: Any) -> bool:
    return (
        info.context
        and info.context.get("skip_validation_on_templates", False)
        and isinstance(value, str)
        and PROJECT_TEMPLATE_START in value
    )


_initial_context: ContextVar[Optional[Dict[str, Any]]] = ContextVar(
    "_init_context_var", default=None
)


@contextmanager
def context(value: Dict[str, Any]) -> Iterator[None]:
    """
    Thread safe context for Pydantic.
    By using `with context()`, you ensure context changes apply
    to the with block only
    """
    token = _initial_context.set(value)
    try:
        yield
    finally:
        _initial_context.reset(token)


class UpdatableModel(BaseModel):
    model_config = ConfigDict(validate_assignment=True, extra="forbid")

    def __init__(self, /, **data: Any) -> None:
        """
        Pydantic provides 2 options to pass in context:
        1) Through `model_validate()` as a second argument.
        2) Through a custom init method and the use of ContextVar

        We decided not to use 1) because it silently stops working
        if someone adds a pass through __init__ to any of the Pydantic models.

        We decided to go with 2) as the safer approach.
        Calling validate_python() in the __init__ is how we can pass context
        on initialization according to Pydantic's documentation:
        https://docs.pydantic.dev/latest/concepts/validators/#using-validation-context-with-basemodel-initialization
        """
        self.__pydantic_validator__.validate_python(
            data,
            self_instance=self,
            context=_initial_context.get(),
        )

    @classmethod
    def _is_entity_type_field(cls, field: Any) -> bool:
        """
        Checks if a field is of type `DiscriminatorField`
        """
        if not isinstance(field, FieldInfo) or not field.json_schema_extra:
            return False

        return (
            "is_discriminator_field" in field.json_schema_extra
            and field.json_schema_extra["is_discriminator_field"]
        )

    @classmethod
    def __init_subclass__(cls, **kwargs):
        """
        This method will collect all the Pydantic annotations for the class
        currently being initialized (any subclass of `UpdatableModel`).

        It will add a field validator wrapper for every Pydantic field
        in order to skip validation when templates are found.

        It will apply this to all Pydantic fields, except for fields
        marked as `DiscriminatorField`. These will be skipped because
        Pydantic does not support validators for discriminator field types.
        """

        super().__init_subclass__(**kwargs)

        field_annotations = {}
        field_values = {}
        # Go through the inheritance classes and collect all the annotations and
        # all the values of the class attributes. We go in reverse order so that
        # values in subclasses overrides values from parent classes in case of field overrides.

        private_attrs = set()
        for class_ in reversed(cls.__mro__):
            class_dict = class_.__dict__
            field_annotations.update(class_dict.get("__annotations__", {}))

            if "model_fields" in class_dict and class_.model_fields:
                # This means the class dict has already been processed by Pydantic
                # All fields should properly be populated in model_fields
                field_values.update(class_.model_fields)
            else:
                # If Pydantic did not process this class yet, get the values from class_dict directly
                field_values.update(class_dict)
            for pa in class_dict.get("__private_attributes__", []):
                private_attrs.add(pa)

        # Add Pydantic validation wrapper around all fields except `DiscriminatorField`s
        for field_name in field_annotations:
            if field_name in private_attrs:
                continue
            field = field_values.get(field_name)
            if not cls._is_entity_type_field(field):
                cls._add_validator(field_name)

    @classmethod
    def _add_validator(cls, field_name: str):
        """
        Adds a Pydantic validator with mode=wrap for the provided `field_name`.
        During validation, this will check if the field is templated (not expanded yet)
        and in that case, it will skip all the remaining Pydantic validation on that field.

        Since this validator is added last, it will skip all the other field validators
        defined in the subclasses when templates are found.

        This logic on templates only applies when context contains `skip_validation_on_templates` flag.
        """

        def validator_skipping_templated_str(cls, value, handler, info: ValidationInfo):
            if _is_templated(info, value):
                return value
            return handler(value)

        setattr(
            cls,
            # Unique name so that subclasses get a unique instance of this validator
            f"_{cls.__module__}.{cls.__name__}_validate_{field_name}",
            field_validator(field_name, mode="wrap")(validator_skipping_templated_str),
        )

    def update_from_dict(self, update_values: Dict[str, Any]):
        """
        Takes a dictionary with values to override.
        If the field type is subclass of a UpdatableModel, its update_from_dict() method is called with
        the value to be set.
        If not, we use simple setattr to set new value.
        Values provided are validated against original restrictions, so it's impossible to overwrite string field with
        integer value etc.
        """
        for field, value in update_values.items():
            if field in self.model_fields.keys():
                if (
                    hasattr(getattr(self, field), "update_from_dict")
                    and field in self.model_fields_set
                ):
                    getattr(self, field).update_from_dict(value)
                else:
                    setattr(self, field, value)
        return self


def DiscriminatorField(*args, **kwargs):  # noqa N802
    """
    Use this type for discriminator fields used for differentiating
    between different entity types.

    When this `DiscriminatorField` is used on a pydantic attribute,
    we will not allow templating on it.
    """
    extra = dict(is_discriminator_field=True)
    return Field(json_schema_extra=extra, *args, **kwargs)


def IdentifierField(*args, **kwargs):  # noqa N802
    return Field(max_length=254, pattern=IDENTIFIER_NO_LENGTH, *args, **kwargs)
