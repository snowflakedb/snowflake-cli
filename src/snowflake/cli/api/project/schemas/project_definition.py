# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass
from types import UnionType
from typing import Any, Dict, List, Optional, Union, get_args, get_origin

from packaging.version import Version
from pydantic import Field, ValidationError, field_validator, model_validator
from pydantic_core.core_schema import ValidationInfo
from snowflake.cli._plugins.nativeapp.entities.application import ApplicationEntityModel
from snowflake.cli._plugins.nativeapp.entities.application_package import (
    ApplicationPackageChildrenTypes,
    ApplicationPackageEntityModel,
)
from snowflake.cli.api.project.errors import SchemaValidationError
from snowflake.cli.api.project.schemas.entities.common import (
    TargetField,
)
from snowflake.cli.api.project.schemas.entities.entities import (
    EntityModel,
    v2_entity_model_types_map,
)
from snowflake.cli.api.project.schemas.updatable_model import UpdatableModel
from snowflake.cli.api.project.schemas.v1.native_app.native_app import (
    NativeApp,
    NativeAppV11,
)
from snowflake.cli.api.project.schemas.v1.snowpark.snowpark import Snowpark
from snowflake.cli.api.project.schemas.v1.streamlit.streamlit import Streamlit
from snowflake.cli.api.utils.types import Context
from typing_extensions import Annotated

AnnotatedEntity = Annotated[EntityModel, Field(discriminator="type")]
scalar = str | int | float | bool


@dataclass
class ProjectProperties:
    """
    This class stores 2 objects representing the snowflake project:

    The project_context object:
    - Used as the context for templating when users reference variables in the project definition file.

    The project_definition object:
    - This is a transformed object type through Pydantic, which has been normalized.
    - This object could have slightly different structure than what the users see in their yaml project definition files.
    - This should be used for the business logic of snow CLI modules.
    """

    project_definition: ProjectDefinition
    project_context: Context


@dataclass
class YamlOverride:
    data: dict | list


class _ProjectDefinitionBase(UpdatableModel):
    def __init__(self, *args, **kwargs):
        try:
            super().__init__(**kwargs)
        except ValidationError as e:
            raise SchemaValidationError(e) from e

    definition_version: Union[str, int] = Field(
        title="Version of the project definition schema, which is currently 1",
    )

    @field_validator("definition_version")
    @classmethod
    def _is_supported_version(cls, version: str) -> str:
        version = str(version)
        version_map = get_version_map()
        if version not in version_map:
            raise ValueError(
                f'Version {version} is not supported. Supported versions: {", ".join(version_map)}'
            )
        return version

    def meets_version_requirement(self, required_version: str) -> bool:
        return Version(self.definition_version) >= Version(required_version)


class DefinitionV10(_ProjectDefinitionBase):
    native_app: Optional[NativeApp] = Field(
        title="Native app definitions for the project", default=None
    )
    snowpark: Optional[Snowpark] = Field(
        title="Snowpark functions and procedures definitions for the project",
        default=None,
    )
    streamlit: Optional[Streamlit] = Field(
        title="Streamlit definitions for the project", default=None
    )


class DefinitionV11(DefinitionV10):
    native_app: Optional[NativeAppV11] = Field(
        title="Native app definitions for the project", default=None
    )
    env: Optional[Dict[str, Union[str, int, bool]]] = Field(
        title="Default environment specification for this project.",
        default=None,
    )


class DefinitionV20(_ProjectDefinitionBase):
    entities: Dict[str, AnnotatedEntity] = Field(
        title="Entity definitions.", default={}
    )
    env: Optional[Dict[str, Union[str, int, bool]]] = Field(
        title="Default environment specification for this project.",
        default=None,
    )
    mixins: Optional[Dict[str, Dict]] = Field(
        title="Mixins to apply to entities",
        default=None,
    )

    @model_validator(mode="after")
    def validate_entities_identifiers(self):
        for key, entity in self.entities.items():
            entity.set_entity_id(key)
            entity.validate_identifier()
        return self

    @field_validator("entities", mode="after")
    @classmethod
    def validate_entities(
        cls, entities: Dict[str, AnnotatedEntity]
    ) -> Dict[str, AnnotatedEntity]:
        for key, entity in entities.items():
            # TODO Automatically detect TargetFields to validate
            if isinstance(entity, list):
                for e in entity:
                    cls._validate_single_entity(e, entities)
            else:
                cls._validate_single_entity(entity, entities)
        return entities

    @classmethod
    def _validate_single_entity(
        cls, entity: EntityModel, entities: Dict[str, AnnotatedEntity]
    ):
        if entity.type == ApplicationEntityModel.get_type():
            if isinstance(entity.from_, TargetField):
                target_key = entity.from_.target
                target_object = entity.from_
                target_type = target_object.get_type()
                cls._validate_target_field(target_key, target_type, entities)
        elif entity.type == ApplicationPackageEntityModel.get_type():
            for child_entity in entity.children:
                target_key = child_entity.target
                cls._validate_target_field(
                    target_key, ApplicationPackageChildrenTypes, entities
                )

    @classmethod
    def _validate_target_field(
        cls, target_key: str, target_type: EntityModel, entities: Dict[str, EntityModel]
    ):
        if target_key not in entities:
            raise ValueError(f"No such target: {target_key}")

        # Validate the target type
        if target_type:
            actual_target_type = entities[target_key].__class__
            if get_origin(target_type) in (Union, UnionType):
                if actual_target_type not in get_args(target_type):
                    expected_types_str = ", ".join(
                        [t.__name__ for t in get_args(target_type)]
                    )
                    raise ValueError(
                        f"Target type mismatch. Expected one of [{expected_types_str}], got {actual_target_type.__name__}"
                    )
            elif target_type is not actual_target_type:
                raise ValueError(
                    f"Target type mismatch. Expected {target_type.__name__}, got {actual_target_type.__name__}"
                )

    @model_validator(mode="before")
    @classmethod
    def apply_mixins(cls, data: Dict, info: ValidationInfo) -> Dict:
        """
        Applies mixins to those entities, whose meta field contains the mixin name.
        """
        if "mixins" not in data or "entities" not in data:
            return data

        duplicated_run = (
            info.context.get("is_duplicated_run", False) if info.context else False
        )
        if not duplicated_run:
            entities = data["entities"]
            for entity_name, entity in entities.items():
                entity_mixins = entity_mixins_to_list(
                    entity.get("meta", {}).get("use_mixins")
                )

                merged_values = cls._merge_mixins_with_entity(
                    entity_id=entity_name,
                    entity=entity,
                    entity_mixins_names=entity_mixins,
                    mixin_defs=data["mixins"],
                )
                entities[entity_name] = merged_values

        return data

    @classmethod
    def _merge_mixins_with_entity(
        cls,
        entity_id: str,
        entity: dict,
        entity_mixins_names: list,
        mixin_defs: dict,
    ) -> dict:
        # Validate mixins
        for mixin_name in entity_mixins_names:
            if mixin_name not in mixin_defs:
                raise ValueError(f"Mixin {mixin_name} not defined")

        # Build object override data from mixins
        data: dict = {}
        for mx_name in entity_mixins_names:
            data = cls._merge_data(data, mixin_defs[mx_name])

        for key, override_value in data.items():
            if key not in get_allowed_fields_for_entity(entity):
                raise ValueError(
                    f"Unsupported key '{key}' for entity {entity_id} of type {entity['type']} "
                )

            entity_value = entity.get(key)
            if (
                entity_value is not None
                and not isinstance(entity_value, YamlOverride)
                and not isinstance(entity_value, type(override_value))
            ):
                raise ValueError(
                    f"Value from mixins for property {key} is of type '{type(override_value).__name__}' "
                    f"while entity {entity_id} expects value of type '{type(entity_value).__name__}'"
                )

        # Apply entity data on top of mixins
        data = cls._merge_data(data, entity)
        return data

    @model_validator(mode="after")
    def validate_dependencies(self):
        """
        Checks if entities listed in depends_on section exist in the project
        """
        missing_dependencies = defaultdict(list)
        for entity_id, entity in self.entities.items():
            if entity.meta:
                for dependency in entity.meta.depends_on:
                    if dependency not in self.entities:
                        missing_dependencies[entity_id].append(dependency)

        if missing_dependencies:
            raise ValueError(_get_missing_dependencies_message(missing_dependencies))

    @classmethod
    def _merge_data(
        cls,
        left: dict | list | scalar | None,
        right: dict | list | scalar | None | YamlOverride,
    ):
        """
        Merges right data into left. Right and left is expected to be of the same type, if not right is returned.
        If left is sequence then missing elements from right are appended.
        If left is dictionary then we update it with data from right. The update is done recursively key by key.
        """
        if isinstance(right, YamlOverride):
            return right.data

        if left is None:
            return right

        # At that point left and right are of the same type
        if isinstance(left, dict) and isinstance(right, dict):
            data = dict(left)
            for key in right:
                data[key] = cls._merge_data(left=data.get(key), right=right[key])
            return data

        if isinstance(left, list) and isinstance(right, list):
            return _unique_extend(left, right)

        if not isinstance(right, type(left)):
            raise ValueError(f"Could not merge {type(right)} and {type(left)}.")

        return right

    def get_entities_by_type(self, entity_type: str):
        return {i: e for i, e in self.entities.items() if e.get_type() == entity_type}


def build_project_definition(**data) -> ProjectDefinition:
    """
    Returns a ProjectDefinition instance with a version matching the provided definition_version value
    """
    if not isinstance(data, dict):
        return
    version = data.get("definition_version")
    version_model = get_version_map().get(str(version))
    if not version or not version_model:
        # Raises a SchemaValidationError
        _ProjectDefinitionBase(**data)
    return version_model(**data)


ProjectDefinitionV1 = Union[DefinitionV10, DefinitionV11]
ProjectDefinitionV2 = DefinitionV20
ProjectDefinition = Union[ProjectDefinitionV1, ProjectDefinitionV2]


def get_version_map():
    version_map = {"1": DefinitionV10, "1.1": DefinitionV11, "2": DefinitionV20}
    return version_map


def entity_mixins_to_list(entity_mixins: Optional[str | List[str]]) -> List[str]:
    """
    Convert an optional string or a list of strings to a list of strings.
    """
    if entity_mixins is None:
        return []
    if isinstance(entity_mixins, str):
        return [entity_mixins]
    return entity_mixins


def get_allowed_fields_for_entity(entity: Dict[str, Any]) -> List[str]:
    """
    Get the allowed fields for the given entity.
    """
    entity_type = entity.get("type")
    if entity_type is None:
        raise ValueError("Entity is missing type declaration.")

    if entity_type not in v2_entity_model_types_map:
        return []

    entity_model = v2_entity_model_types_map[entity_type]
    return entity_model.model_fields


def _unique_extend(list_a: List, list_b: List) -> List:
    new_list = list(list_a)
    for item in list_b:
        if all(item != x for x in list_a):
            new_list.append(item)
    return new_list


def _get_missing_dependencies_message(
    missing_dependencies: Dict[str, List[str]]
) -> str:
    missing_dependencies_message = []
    for entity_id, dependencies in missing_dependencies.items():
        missing_dependencies_message.append(
            f"\n Entity {entity_id} depends on non-existing entities: {', '.join(dependencies)}"
        )
    return "".join(missing_dependencies_message)
