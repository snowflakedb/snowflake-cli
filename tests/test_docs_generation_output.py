# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from pathlib import Path
from textwrap import dedent
from unittest import mock

from click import Command
from pydantic.json_schema import GenerateJsonSchema, model_json_schema
from snowflake.cli.api.project.schemas.project_definition import DefinitionV11


@mock.patch(
    "snowflake.cli._app.dev.docs.project_definition_generate_json_schema.ProjectDefinitionGenerateJsonSchema.generate"
)
def test_definition_file_format_generated_from_json(mock_generate, runner, temp_dir):
    property1 = {
        "path": "propertyA",
        "title": "Title of property A",
        "indents": 0,
        "item_index": 0,
        "required": False,
        "name": "propertyA",
        "add_types": False,
        "types": "",
    }

    property2 = {
        "path": "propertyA.propertyB",
        "title": "Title of property B",
        "description": "Description of property B",
        "indents": 1,
        "item_index": 1,
        "required": True,
        "name": "propertyB",
        "add_types": True,
        "types": "string",
    }
    result = {
        "result": [
            {
                "properties": [property1, property2],
                "title": "SECTION_TITLE",
                "name": "section_demo",
            }
        ]
    }

    mock_generate.return_value = result

    runner.invoke(["--docs"])
    project_definition_path = (
        Path(temp_dir)
        / "gen_docs"
        / "project_definition"
        / "definition_section_demo.txt"
    )

    assert project_definition_path.read_text() == dedent(
        """\
SECTION_TITLE

Project definition structure
===============================================================================

.. code-block::
    
    
  propertyA: 
    - propertyB: <string>


Project definition properties
===============================================================================
The following table describes the project definition properties.
    
.. list-table:: Project definition properties
  :widths: 30 70
  :header-rows: 1
    
  * - Property
    - Definition

  * - **propertyA**

      *Optional*

    - Title of property A
 
  * - **propertyA.propertyB**

      *Required*, *string*

    - Title of property B
    
      Description of property B

"""
    )


def test_files_generated_for_each_optional_project_definition_property(
    runner, temp_dir
):
    runner.invoke(["--docs"])
    project_definition_path = Path(temp_dir) / "gen_docs" / "project_definition"
    errors = []

    model_json = model_json_schema(DefinitionV11, schema_generator=GenerateJsonSchema)
    for property_name in model_json["properties"]:
        if property_name in model_json["required"]:
            continue
        if not (project_definition_path / f"definition_{property_name}.txt").exists():
            errors.append(f"Section `{property_name}` was not properly generated")

    assert len(errors) == 0, "\n".join(errors)


def test_all_commands_have_generated_files(runner, temp_dir, get_click_context):
    runner.invoke(["--docs"])

    # invoke help command to populate app context (plugins registration)
    runner.invoke(["--help"])

    commands_path = Path(temp_dir) / "gen_docs" / "commands"

    errors = []

    def _check(command: Command, directory_path: Path, command_path=None):
        if command_path is None:
            command_path = []
        if getattr(command, "hidden", False):
            return
        if hasattr(command, "commands"):
            for command_name, command_info in command.commands.items():
                new_directory_path = (
                    directory_path / command.name
                    if command.name != "default"
                    else directory_path
                )
                _check(command_info, new_directory_path, [*command_path, command_name])
        else:
            if not (directory_path / f"usage-{command.name}.txt").exists():
                errors.append(
                    f"Command `{' '.join(command_path)}` documentation was not properly generated"
                )

    _check(get_click_context().command, commands_path)

    assert len(errors) == 0, "\n".join(errors)


def test_flags_have_default_values(runner, temp_dir, snapshot):
    runner.invoke(["--docs"])

    # cortex complete checks:
    # "Default: False" case
    # "--diag-log-path" flag, with tempdir path as default value
    example_generated_file = (
        Path(temp_dir) / "gen_docs" / "commands" / "cortex" / "usage-complete.txt"
    )
    assert example_generated_file.exists()
    assert example_generated_file.read_text() == snapshot
