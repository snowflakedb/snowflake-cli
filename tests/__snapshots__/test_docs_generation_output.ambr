# serializer version: 1
# name: test_flags_have_default_values
  '''
  Given a prompt, the command generates a response using your choice of language model. In the simplest use case, the prompt is a single string. You may also provide a JSON file with conversation history including multiple prompts and responses for interactive chat-style usage.
  
  Syntax
  ===============================================================================
  
  .. code-block:: console
  
    snow cortex complete
      <text>
      --model <model>
      --file <file>
      --connection <connection>
      --host <host>
      --port <port>
      --account <account>
      --user <user>
      --password <password>
      --authenticator <authenticator>
      --private-key-file <private_key_file>
      --token-file-path <token_file_path>
      --database <database>
      --schema <schema>
      --role <role>
      --warehouse <warehouse>
      --temporary-connection
      --mfa-passcode <mfa_passcode>
      --enable-diag
      --diag-log-path <diag_log_path>
      --diag-allowlist-path <diag_allowlist_path>
      --format <format>
      --verbose
      --debug
      --silent
  
  Arguments
  ===============================================================================
  
  :samp:`{text}`
    Prompt to be used to generate a completion. Cannot be combined with --file option.
  
  Options
  ===============================================================================
  
  :samp:`--model {TEXT}`
    String specifying the model to be used. Default: snowflake-arctic.
  
  :samp:`--file {FILE}`
    JSON file containing conversation history to be used to generate a completion. Cannot be combined with TEXT argument.
  
  :samp:`--connection, -c, --environment {TEXT}`
    Name of the connection, as defined in your `config.toml` file. Default: `default`.
  
  :samp:`--host {TEXT}`
    Host address for the connection. Overrides the value specified for the connection.
  
  :samp:`--port {INTEGER}`
    Port for the connection. Overrides the value specified for the connection.
  
  :samp:`--account, --accountname {TEXT}`
    Name assigned to your Snowflake account. Overrides the value specified for the connection.
  
  :samp:`--user, --username {TEXT}`
    Username to connect to Snowflake. Overrides the value specified for the connection.
  
  :samp:`--password {TEXT}`
    Snowflake password. Overrides the value specified for the connection.
  
  :samp:`--authenticator {TEXT}`
    Snowflake authenticator. Overrides the value specified for the connection.
  
  :samp:`--private-key-file, --private-key-path {TEXT}`
    Snowflake private key file path. Overrides the value specified for the connection.
  
  :samp:`--token-file-path {TEXT}`
    Path to file with an OAuth token that should be used when connecting to Snowflake.
  
  :samp:`--database, --dbname {TEXT}`
    Database to use. Overrides the value specified for the connection.
  
  :samp:`--schema, --schemaname {TEXT}`
    Database schema to use. Overrides the value specified for the connection.
  
  :samp:`--role, --rolename {TEXT}`
    Role to use. Overrides the value specified for the connection.
  
  :samp:`--warehouse {TEXT}`
    Warehouse to use. Overrides the value specified for the connection.
  
  :samp:`--temporary-connection, -x`
    Uses a connection defined with command line parameters, instead of one defined in config. Default: False.
  
  :samp:`--mfa-passcode {TEXT}`
    Token to use for multi-factor authentication (MFA).
  
  :samp:`--enable-diag`
    Whether to generate a connection diagnostic report. Default: False.
  
  :samp:`--diag-log-path {TEXT}`
    Path for the generated report. Defaults to system temporary directory. Default: <system_temporary_directory>.
  
  :samp:`--diag-allowlist-path {TEXT}`
    Path to a JSON file containing allowlist parameters.
  
  :samp:`--format [TABLE|JSON]`
    Specifies the output format. Default: TABLE.
  
  :samp:`--verbose, -v`
    Displays log entries for log levels `info` and higher. Default: False.
  
  :samp:`--debug`
    Displays log entries for log levels `debug` and higher; debug logs contain additional information. Default: False.
  
  :samp:`--silent`
    Turns off intermediate output to console. Default: False.
  
  '''
# ---
